### ISSUE TITLE
DOCS: Extra sections on the performance page

### ISSUE THREAD

**Issue by JHopeCollins** [firedrakeproject/firedrake#3813]
Following on from a recent discussion on slack, it would be useful to add some more sections to the manual page on performance. Some possible topics include:

1. Using `Constant` or `Real` instead of `float` to avoid recompiling kernels every time the value changes.
2. Setting `constant_jacobian=True` if the coefficients in a `LinearVariationalProblem` don't change between solves (or if they don't change every time, and use the `invalidate_jacobian()` method).
3. Consider changing to an iterative method instead of the default LU once you scale up past little test problems.

**Where to find**
https://www.firedrakeproject.org/optimising.html

**Additional info**
Flamegraphs of before/after each change would be nice, and adding them to the existing "Calling `solve` repeatedly" section.

**Comment by pbrubeck** [comment]
Regarding 2. aren't we clever enough to do this automatically if `coefficient.dat.dat_version` did not change for every `coefficient` in `form.coefficients()` since the last time we assembled?

**Comment by JHopeCollins** [comment]
Maybe, but we don't. You also need to look at all the `Constants` in the form not just the dats. There may well be other things you'd need to check to do it reliably.

Combined with defaulting to LU, having `constant_jacobian=False` as the default can be a big performance hit so it should be in this manual section.

If we start doing it automagically then we can take that topic out.

**Comment by dham** [comment]
I agree with Josh. Automating constant Jacobian detection looks like significant work to do robustly. Not saying we shouldn't, but it's not trivial.

**Comment by dham** [comment]
@koki thinks maybe this can easily and robustly done in TwoFormAssembler by inspecting the Parloop arguments.
